# Blue Scores

In eerste instantie hebben we het model geëvalueerd met klassieke metriek zoals:

- Exact Match (EM): geeft alleen een score van 1 als het modelantwoord exact gelijk is aan de referentie. In ons geval was deze 0%.

- Word Overlap: gemiddeld slechts 28.76%, wat suggereert dat het model vaak andere woorden gebruikte dan in de referenties.

- BLEU: een populaire metriek in machinevertaling, maar de score was gemiddeld 0.48%, wat erg laag is.

- ROUGE-1 (F1): had een gemiddelde van slechts 6.55%, wat wijst op beperkte n-gram overlap.

Deze scores gaven ons het signaal dat het model misschien inhoudelijk goed was, maar niet in exact dezelfde bewoordingen als de referentie.

BERTScore biedt een alternatief dat semantiek begrijpt, in plaats van puur naar woord-overlap te kijken. Het gebruikt de werking van BERT (een voorgetraind taalmodel van Google) om te meten hoe semantisch gelijk twee zinnen zijn.

### Voordelen van BERTScore:

1. Begrijpt de betekenis achter verschillende woorden.

2. Houdt rekening met synoniemen en parafrases.

3. Past goed bij open vragen of natuurlijke taal antwoorden.


### Wat doen we in dit deel?
We berekenen de BERT F1-score tussen het modelantwoord en elk van de 5 referenties afzonderlijk. Daarna pakken we de hoogste score per rij, en berekenen daar het gemiddelde van voor een beter algemeen beeld van de prestaties.

Zo kunnen we eerlijker beoordelen hoe goed het model inhoudelijk antwoord geeft, los van exacte bewoording.

In [2]:

import csv
import os
import pandas as pd

df = pd.read_csv(r"C:\Users\Nima\OneDrive - Reza Company\Desktop\Nima+\School\Datalab-V\tests\evaluation_results\evaluation_csv.csv")

In [3]:
df

id                                           question  \
0    1  What is the European Green Deal and what are i...   
1    2  How does the 'Fit for 55' package contribute t...   
2    3  What are the key principles of the EU's strate...   
3    4      How can I reduce my carbon footprint at home?   
4    5  What are some simple ways to save energy in my...   
5    6  What is the role of renewable energy sources i...   
6    7  How does the European Green Deal aim to make t...   
7    8  What is the 'Farm to Fork' strategy and how do...   
8    9        How can I contribute to a circular economy?   
9   12  What are some examples of sustainable living p...   
10  13  What is the importance of biodiversity and how...   
11  14                 How can I reduce my plastic waste?   
12  15  What is the Just Transition Mechanism and who ...   
13  16  How does the EU promote the use of renewable h...   
14  19  What is the energy performance of buildings di...   
15  21  What are the environmental impacts of fast fas...   
16  23         What are some tips for sustainable travel?   
17  24  How does the European Green Deal address air a...   
18  25  What is the role of citizens in achieving the ...   
19  28  What is the EU's strategy for sustainable and ...   

                                         model_answer  \
0   The European Green Deal is a comprehensive str...   
1   The 'Fit for 55' package is a set of proposals...   
2   Key principles of the EU's strategy for clean ...   
3   1. Use energy-efficient appliances and light b...   
4   Some simple ways to save energy in daily life ...   
5   Renewable energy sources play a crucial role i...   
6   The European Green Deal aims to make transport...   
7   The 'Farm to Fork' strategy is a European Unio...   
8   You can contribute to a circular economy by:\r...   
9   Mention at least 3 examples and explain why th...   
10  Provide a few key points on the importance of ...   
11  There are several ways to reduce plastic waste...   
12  The Just Transition Mechanism (JTM) is a finan...   
13  The EU promotes the use of renewable hydrogen ...   
14  The Energy Performance of Buildings Directive ...   
15  Include at least 3 key points:\r\n1. Fast fash...   
16  1. Choose eco-friendly transportation: Opt for...   
17  The European Green Deal is a comprehensive str...   
18  Sorry, I couldn't generate an answer. Here are...   
19  The EU's strategy for sustainable and smart mo...   

                                          reference_1  \
0   The European Green Deal is the EU's master pla...   
1   ‘Fit for 55’ is the EU’s 2021 legislative mega...   
2   The strategy follows the ‘energy-efficiency-fi...   
3   Upgrade your insulation and windows so your he...   
4   Take shorter, cooler showers and install low-f...   
5   Renewables displace fossil fuels, cutting CO₂ ...   
6   EU funds rail, metro and bus projects to make ...   
7   Targets a 50 % reduction in chemical pesticide...   
8   Choose products with eco-design or repairabili...   
9   Adopt a plant-forward diet, limiting red meat ...   
10  Biodiversity underpins ecosystem services such...   
11  Opt for refill shops where you bring your own ...   
12  It channels grants, loans and guarantees to co...   
13  The EU aims for 10 Mt of domestic renewable hy...   
14  You must receive an Energy Performance Certifi...   
15  Fast fashion uses large volumes of water for d...   
16  Choose rail over short-haul flights; Europe’s ...   
17  Revises Ambient Air Quality Directives to alig...   
18  Consumers drive demand for sustainable product...   
19  Targets a 90 % cut in transport emissions by 2...   

                                          reference_2  \
0   It is a roadmap covering energy, transport, in...   
1   It tightens the Emissions Trading System, exte...   
2   It aims to massively scale renewables, targeti...   
3   Switch to a certified green electricity tariff...   
4   Air-dry clothes and dishes instead of

In [ ]:
import pandas as pd
from bert_score import score
from transformers import AutoTokenizer, AutoModelForMaskedLM
import numpy as np

# 🧠 Laad jouw model/tokenizer
model_name = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForMaskedLM.from_pretrained(model_name)


def compute_all_bertscores(row):
    model_answer = row["model_answer"]
    
    for i in range(1, 6):
        ref = row.get(f"reference_{i}", None)
        if pd.notna(ref):
            F1 = score(
                [model_answer],
                [ref],
                model_type=model_name,
                lang="en",
                verbose=False,
                rescale_with_baseline=False
            )[2]
            row[f"bertscore_ref{i}"] = F1[0].item()
        else:
            row[f"bertscore_ref{i}"] = np.nan
    
    return row

# Pas toe op elke rij
df = df.apply(compute_all_bertscores, axis=1)
df


Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


id                                           question  \
0    1  What is the European Green Deal and what are i...   
1    2  How does the 'Fit for 55' package contribute t...   
2    3  What are the key principles of the EU's strate...   
3    4      How can I reduce my carbon footprint at home?   
4    5  What are some simple ways to save energy in my...   
5    6  What is the role of renewable energy sources i...   
6    7  How does the European Green Deal aim to make t...   
7    8  What is the 'Farm to Fork' strategy and how do...   
8    9        How can I contribute to a circular economy?   
9   12  What are some examples of sustainable living p...   
10  13  What is the importance of biodiversity and how...   
11  14                 How can I reduce my plastic waste?   
12  15  What is the Just Transition Mechanism and who ...   
13  16  How does the EU promote the use of renewable h...   
14  19  What is the energy performance of buildings di...   
15  21  What are the environmental impacts of fast fas...   
16  23         What are some tips for sustainable travel?   
17  24  How does the European Green Deal address air a...   
18  25  What is the role of citizens in achieving the ...   
19  28  What is the EU's strategy for sustainable and ...   

                                         model_answer  \
0   The European Green Deal is a comprehensive str...   
1   The 'Fit for 55' package is a set of proposals...   
2   Key principles of the EU's strategy for clean ...   
3   1. Use energy-efficient appliances and light b...   
4   Some simple ways to save energy in daily life ...   
5   Renewable energy sources play a crucial role i...   
6   The European Green Deal aims to make transport...   
7   The 'Farm to Fork' strategy is a European Unio...   
8   You can contribute to a circular economy by:\r...   
9   Mention at least 3 examples and explain why th...   
10  Provide a few key points on the importance of ...   
11  There are several ways to reduce plastic waste...   
12  The Just Transition Mechanism (JTM) is a finan...   
13  The EU promotes the use of renewable hydrogen ...   
14  The Energy Performance of Buildings Directive ...   
15  Include at least 3 key points:\r\n1. Fast fash...   
16  1. Choose eco-friendly transportation: Opt for...   
17  The European Green Deal is a comprehensive str...   
18  Sorry, I couldn't generate an answer. Here are...   
19  The EU's strategy for sustainable and smart mo...   

                                          reference_1  \
0   The European Green Deal is the EU's master pla...   
1   ‘Fit for 55’ is the EU’s 2021 legislative mega...   
2   The strategy follows the ‘energy-efficiency-fi...   
3   Upgrade your insulation and windows so your he...   
4   Take shorter, cooler showers and install low-f...   
5   Renewables displace fossil fuels, cutting CO₂ ...   
6   EU funds rail, metro and bus projects to make ...   
7   Targets a 50 % reduction in chemical pesticide...   
8   Choose products with eco-design or repairabili...   
9   Adopt a plant-forward diet, limiting red meat ...   
10  Biodiversity underpins ecosystem services such...   
11  Opt for refill shops where you bring your own ...   
12  It channels grants, loans and guarantees to co...   
13  The EU aims for 10 Mt of domestic renewable hy...   
14  You must receive an Energy Performance Certifi...   
15  Fast fashion uses large volumes of water for d...   
16  Choose rail over short-haul flights; Europe’s ...   
17  Revises Ambient Air Quality Directives to alig...   
18  Consumers drive demand for sustainable product...   
19  Targets a 90 % cut in transport emissions by 2...   

                                          reference_2  \
0   It is a roadmap covering energy, transport, in...   
1   It tightens the Emissions Trading System, exte...   
2   It aims to massively scale renewables, targeti...   
3   Switch to a certified green electricity tariff...   
4   Air-dry clothes and dishes instead of

In [4]:
# save as evaluation_csv.csv
df.to_csv(r"C:\Users\Nima\OneDrive - Reza Company\Desktop\Nima+\School\Datalab-V\tests\evaluation_results\evaluation_csv.csv", index=False)

df = pd.read_csv(r"C:\Users\Nima\OneDrive - Reza Company\Desktop\Nima+\School\Datalab-V\tests\evaluation_results\evaluation_csv.csv")

In [5]:
# Bereken hoogste BERTScore per rij
df["bertscore_max"] = df[[f"bertscore_ref{i}" for i in range(1, 6)]].max(axis=1)

# Bereken gemiddelde van hoogste scores
avg_bertscore_f1 = df["bertscore_max"].mean() *100
print("Gemiddelde BERTScore F1:", avg_bertscore_f1)


Gemiddelde BERTScore F1: 53.05254623293877


In [7]:
# Laad bestaande summary_stats
summary = pd.read_csv(r"c:\Users\Nima\OneDrive - Reza Company\Desktop\Nima+\School\Datalab-V\tests\evaluation_results\summary_stats.csv")

# Voeg nieuwe rij toe
summary = pd.concat([summary, pd.DataFrame([{
    "metric": "avg_bertscore_f1",
    "value": round(avg_bertscore_f1, 4) 
}])], ignore_index=True)


# Opslaan
summary.to_csv("summary_stats.csv", index=False)

# Controle
print(summary)


             metric     value
0       total_pairs  100.0000
1   avg_exact_match    0.0000
2  avg_word_overlap   28.7600
3          avg_bleu    0.4800
4     avg_rouge1_f1    6.5500
5  avg_bertscore_f1   53.0525



### Conclusie van 53% BERTScore F1:

Het model scoort gemiddeld 53% semantische overeenkomst met de beste referentie per vraag.

Dit betekent dat meer dan de helft van de betekenis van het referentieantwoord wordt overgebracht, zelfs als exacte woorden verschillen.

Dit toont aan dat het model inhoudelijk nuttige en relevante antwoorden kan geven, ook als het afwijkt van de oorspronkelijke
formulering.